Add support for Playwright Component and E2E Testing

n05la3 opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/quasarframework/quasar-testing/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

<!-- PULL REQUEST TEMPLATE -->
<!-- (Update "[ ]" to "[x]" to check a box) -->

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] New test runner
- [ ] Documentation
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**If you are adding a new test runner, have you...?** (check all)

- [ ] Created an issue first?
- [ ] Registered it in `/packages/base/runners.json`?
- [ ] Added it to `/README.md`?
- [ ] Included one test that runs `baseline.spec.vue`?
- [ ] Added and updated documentation?
- [ ] Included a recipe folder with properly building quasar project?

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [ ] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] It's submitted to the `dev` branch and _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] It's been tested on Windows
- [ ] It's been tested on Linux
- [ ] It's been tested on MacOS
- [ ] Any necessary documentation has been added or updated [in the docs](https://github.com/quasarframework/quasar-framework.org/tree/master/source) (for faster update click on "Suggest an edit on GitHub" at bottom of page) or explained in the PR's description.

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

n05la3 committed:
feat(playwright): add prompts

n05la3 committed:
feat(playwright): initial setup

n05la3 committed:
feat(playwright): setup esm and ejs build

n05la3 committed:
chore(playwright): setup eslint and husky

n05la3 committed:
feat(playwright): set compatible Quasar version and enforce server port

n05la3 committed:
feat(playwright): add scripts to run in watch mode

n05la3 committed:
feat(playwright): implement AE installation

n05la3 committed:
feat(playwright): add files to `.gitignore` file during installation

n05la3 committed:
feat(client): control the installation instead of depending on Playwright

n05la3 committed:
feat(playwright): add sample workflow file

n05la3 committed:
feat(playwright): add Playwright configs for component and e2e testing

n05la3 committed:
feat(playwright): add color utility

n05la3 committed:
feat(playwright): add portal utils

n05la3 committed:
fix(playwright): add default port

n05la3 committed:
feat(playwright): add color assertions utils

n05la3 committed:
feat(playwright): add test route utils

n05la3 committed:
feat(playwright): export utils for library use

n05la3 committed:
feat(playwright): add components for test

n05la3 committed:
feat(playwright): add more component templates and config files

n05la3 committed:
feat(playwright): add js and ts templates

n05la3 committed:
feat(playwright): setup app-vite playwright test app

n05la3 committed:
feat(playwright): add component tests

yusufkandemir commented on packages/e2e-playwright/package.json:
```diff
@@ -0,0 +1,62 @@
+  "license": "MIT",
+  "type": "module",
+  "scripts": {
+    "build:esm": "tsc -p tsconfig.json",
+    "build:cjs": "tsc -p tsconfig-cjs.json",
+    "build:types": "tsc --emitDeclarationOnly --outDir dist/types",
+    "build": "rimraf ./dist && yarn build:esm && yarn build:cjs && yarn build:types",
```
I would recommend using https://github.com/unjs/unbuild

n05la3 replied:
Alright, thanks. Will check this out.

n05la3 replied:
Checked out unbuild and went with it in the end. It made it easier to support cjs and esm builds.

yusufkandemir commented on packages/e2e-playwright/package.json:
```diff
@@ -0,0 +1,62 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/n05la3/testing-e2e-playwright"
```
```suggestion
    "url": "https://github.com/quasarframework/quasar-testing",
    "directory": "packages/e2e-playwright"
```

yusufkandemir commented on packages/e2e-playwright/package.json:
```diff
@@ -0,0 +1,62 @@
+  },
+  "main": "dist/cjs/main.js",
+  "module": "dist/esm/main.js",
+  "bugs": "",
+  "homepage": "",
```
```suggestion
  "bugs": "https://github.com/quasarframework/quasar-testing/issues",
  "homepage": "https://github.com/quasarframework/quasar-testing/tree/dev/packages/e2e-playwright",
```

yusufkandemir commented on packages/e2e-playwright/src/prompts.js:
```diff
@@ -0,0 +1,35 @@
+      required: true,
+      default: 'y/N',
+      message: 'Do you want to enable code coverage?',
+    },
```
There is no type `text`. It works because it falls back to the default: `input`
Let's use `confirm` type for y/n questions to improve both UX and DX.

```suggestion
    {
      name: 'port',
      type: 'number',
      required: true,
      default: enforcedDevServerPort,
      message: 'Set the port from which to serve the app:',
    },
    {
      name: 'githubWorkflow',
      type: 'confirm',
      required: true,
      default: false,
      message: 'Do you want to add a GitHub workflow for your CI?',
    },
    // TODO: allow selecting which browsers to install
    {
      name: 'installBrowsers',
      type: 'confirm',
      required: true,
      default: false,
      message:
        "Install Playwright browsers (can be done manually via 'npx exec playwright install')?",
    },
    {
      name: 'enableCodeCoverage',
      type: 'confirm',
      required: true,
      default: false,
      message: 'Do you want to enable code coverage?',
    },
```

yusufkandemir commented on packages/e2e-playwright/src/templates/javascript/playwright-ct.config.js:
```diff
@@ -0,0 +1,83 @@
+    /* Port to use for Playwright component endpoint. */
+    ctPort: 3100,
+
+    ctViteConfig: {
```
I would say we should try passing the whole config with the `@quasar/app-vite/lib/testing.js`, or a slimmed-down version of it, instead of just the `vue` and `quasar` plugins. This is especially important for allowing the use of aliases, e.g. `import x from 'src/foo'`

n05la3 replied:
Did you mean `@quasar/app-vite/lib/config-tools.js`? Could you please check if you provided the wrong path? I checked and did not find the `testing.js` in that path.

yusufkandemir replied:
It's imported as `@quasar/app-vite/lib/testing.js`, but live inside `app-vite/exports/testing/testing.js`
https://github.com/quasarframework/quasar/blob/259a400d258c26fc5b7fd068e9247e189f574af4/app-vite/package.json#L49-L51
https://github.com/quasarframework/quasar/blob/dev/app-vite/exports/testing/testing.js

yusufkandemir commented on packages/e2e-playwright/src/templates/test-watcher.ts:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Playwright does not yet offer a means to run the tests in a watch mode.
```
It is possible to pass this env variable to achieve it: `PWTEST_WATCH=1`
https://github.com/microsoft/playwright/issues/21960

n05la3 replied:
Alright, thanks. I already got rid of that file. It was part of an early iteration. There's also a button when running in UI mode that can be used to activate/deactivate watch mode. I'll push the changes soon.

yusufkandemir commented on packages/e2e-playwright/src/install.js:
```diff
@@ -0,0 +1,174 @@
+
+    if (await api.hasLint()) {
+      api.onExitLog(
+        'Check out https://github.com/quasarframework/quasar-testing/tree/dev/packages/e2e-cypress to see how to add proper Cypress linting configuration to your project.',
```
```suggestion
        'Check out https://github.com/quasarframework/quasar-testing/tree/dev/packages/e2e-playwright to see how to add proper Playwright linting configuration to your project.',
```

yusufkandemir commented on packages/e2e-playwright/src/install.js:
```diff
@@ -0,0 +1,174 @@
+      },
+    };
+
+    // Playwirght only offers native support for component testing with Vite
```
```suggestion
    // Playwright only offers native support for component testing with Vite
```

yusufkandemir reviewed (changes requested):
I know the PR is not ready for review, but wanted to leave an early review

mtdvlpr commented:
@yusufkandemir, could you also review #390 again? A lot of people are waiting on it so Vitest 3 could finally be supported.

yusufkandemir commented:
@mtdvlpr that's ready to be merged, internally known. But I don't manage the releases, so there's nothing more I can do.
/cc @IlCallo

n05la3 committed:
fix(playwright): adjust how prompts are handled

IlCallo commented:
@n05la3 please rebase this PR on dev branch and address the issues Yusuf mentioned

n05la3 committed:
feat(playwright): add support for component testing to component config

n05la3 committed:
feat(playwright): add code coverage flag to e2e template

n05la3 committed:
refactor(playwright): remove test-route.ts utility function

- Let's prefer the built in route test

n05la3 committed:
refactor(playwright): rename color assertions

n05la3 committed:
feat(playwright): add component test fixture

n05la3 committed:
refactor(playwright): move tests to base template

n05la3 committed:
fix(playwright): fix bug preventing prompt in new stable app-vite versions

- This bug was caused wrong export definition in package.json which prevented the prompt from being imported correctly

n05la3 committed:
feat(playwright): many random improvement for code coverage support

n05la3 committed:
feat(playwright): init empty project and setup playwright

n05la3 committed:
feat(playwright): sync project

n05la3 committed:
refactor(playwright): adjust the nyc config

n05la3 committed:
fix(playwright): fix component fixture typing

n05la3 committed:
refactor(plawright): improve templates

n05la3 committed:
feat(playwright): add flag to control whether to enable coverage or not

n05la3 committed:
feat(playwright): prefer unbuild to tsc for building the extension

- Unbuild offers a better way to generate a build that works well with commonjs and module. It can correcly name all module files as .mjs and common as .cjs, this makes it easy for consumers to import

n05la3 committed:
chore(playwright): sync test project

n05la3 committed:
feat(playwright): sync test apps

n05la3 committed:
refactor(playwright): minor utility improvement

n05la3 committed:
manor minor fixes, resync/update test apps
